### ISSUE TITLE
Enforce uniqueness when using Oban.insert_all()

### ISSUE THREAD

**Issue by bartblast** [sorentwo/oban#388]
At the moment enforcing uniqueness works only for single message - Oban.insert().

**Comment by sorentwo** [comment]
@bartblast This is a little trickier than it sounds. Uniqueness is enforced using two mechanisms:

1. Advisory locks (using a key built off the worker/queue/state/args)
2. Existence search (using a dynamic query built off the worker/queue/state/args)

Enforcing uniqueness for `insert_all` requires a few changes:

1. Advisory locks are before any insertion begins for every job with a `unique` config
2. Performing an existence search for those `unique` jobs before insertion, then use the results to prevent any duplicate inserts by comparing existing records with the passed changesets

In short, it is doable, but difficult. You didn't indicate whether this was intended as a bug report or a feature request, but I'm treating it as the latter.

**Comment by bartblast** [comment]
@sorentwo Yeah, it's a feature request. Any chance it is added to the roadmap?

**Comment by bartblast** [comment]
PS: It would be helpful if this problem was mentioned in the docs, because it is not obvious how it works (I encountered the problem during tests, although I read the whole docs).

**Comment by sorentwo** [comment]
> Any chance it is added to the roadmap?

Indeed, it is on the roadmap. I'm leaving this issue open to keep track of it.

> PS: It would be helpful if this problem was mentioned in the docs, because it is not obvious how it works (I encountered the problem during tests, although I read the whole docs).

It is mentioned in the docs for `insert_all/2`:

> Insertion respects prefix and log settings, but it _does not use per-job unique configuration_. You must use insert/2,4 or insert!/2 for per-job unique support.

It could be a little clearer, and should also be included in `insert_all/4`.

**Comment by bartblast** [comment]
Got it, thanks!

**Comment by dsdshcym** [comment]
Hi @sorentwo!

> 2. Performing an existence search for those unique jobs before insertion, then use the results to prevent any duplicate inserts by comparing existing records with the passed changesets

Can we leverage the `:on_conflict` option to perform updates when conflicting?
I think this can achieve the expected result without adding any complexity.
WDYT?

**Comment by sorentwo** [comment]
> Can we leverage the `:on_conflict` option to perform updates when conflicting?
> I think this can achieve the expected result without adding any complexity.
> WDYT?

@dsdshcym The `insert_all` function already uses `on_conflict` and if you want to create custom unique indexes it will work just fine. The goal is to mimic what `insert` does, which doesn't require any index management.

Side note: this is a "nice to have" change for consistency, but it hasn't been a priority because you can easily use a multi to insert jobs in bulk.

**Comment by dsdshcym** [comment]
Got it, thanks for the explanation! 👍

**Comment by romariolopezc** [comment]
Hello. For some months I've been dealing with duplicated jobs and I couldn't find why. I even increased the unique period 😅 

I re-read the documentation and found that `insert_all`: `Insertion respects prefix and log settings, but it does not use per-job unique configuration. You must use insert/2,4 or [insert!/2](https://hexdocs.pm/oban/Oban.html#insert!/2) for per-job unique support.`

I think the docs can be improved to make this bigger / bolder or with a "disclaimer" in the same fn.

What do you think @sorentwo ?

**Comment by sorentwo** [comment]
Until I can get to support for unique with insert_all, a bolder disclaimer sounds good. PR welcome!

**Comment by sorentwo** [comment]
The SmartEngine's version of `insert_all` now supports unique, replace, and auto-batching. Available in the next Pro release. 🚀

**Comment by Paryz** [comment]
hey @sorentwo is there any date related to this next release? I won't lie it would help me a lot with one of our features 😅

**Comment by sorentwo** [comment]
The next Oban and Pro release will be within the next two weeks. (Note that this is specifically a Pro feature)

**Comment by Paryz** [comment]
yeah, company I work for owns the pro license 😄
